kernel/byterun/coq_uint63_native.h

```diff
@@ -46,17 +55,26 @@
   if (uint63_lsl_y__ < (uint64_t) 127) \
     accu = (value)((((uint64_t)(x)-1) << uint63_of_value(uint63_lsl_y__)) | 1); \
   else \
-    accu = uint63_zero; \
+    accu = uint63_zero(); \
   }while(0)
 #define Uint63_lsr(x,y) do{ \
   value uint63_lsl_y__ = (y); \
   if (uint63_lsl_y__ < (uint64_t) 127) \
     accu = (value)(((uint64_t)(x) >> uint63_of_value(uint63_lsl_y__)) | 1); \
   else \
-    accu = uint63_zero; \
+    accu = uint63_zero(); \
+  }while(0)
+/* FIXME this definition is the same as lsr for now */
+#define Uint63_asr(x,y) do{ \
+    value uint63_asr_y__ = (y); \
+    if (uint63_asr_y__ < (uint64_t) 127) \
+      accu = (value)(((uint64_t)(x) >> uint63_of_value(uint63_asr_y__)) | 1); \
+    else \
+      accu = uint63_zero(); \
   }while(0)
 #define Uint63_lsl1(x) (accu = (value)((((uint64_t)(x)-1) << 1) +1))
 #define Uint63_lsr1(x) (accu = (value)(((uint64_t)(x) >> 1) |1))
+#define Uint63_asr1(x) (accu = (value)(((uint64_t)(x) >> 1) |1)) /* FIXME this is the same as asr1 */
```

Comment:
I have no idea how this works nor how to adapt it to do an arithmetic shift instead of a logical one. Can someone help me please?

Reply:
OCaml GC uses the least significant bit to distinguish integer values from adresses (pointers). Adresses being aligned, they always end up with 0 so integers are shifted left by 1 bit to put a 1 at the end (x is represented by 2x+1). So arithmetic shift right of x should be represented by (arithmetic shift right of x) | 1 (to ensure the least significant bit is forced to 1).

Reply:
So, in practice replacing `uint64_t` by `int64_t` should be fine.

Reply:
I would go for the same trick we used for signed division. First you do a left shift of `(uint64_t)x` by one, then you cast to `int64_t`, then you do a right shift by `uint63_of_value(uint63_asr_y__) + 1`, then you mask to clear the sign bit.

Reply:
> So, in practice replacing uint64_t by int64_t should be fine.

Oops, I had not seen your message. Your solution is much better.